### PR TITLE
hnsw snapshotting

### PR DESCRIPTION
**Pull request by jeroiraz**
### What's being changed:

This PR introduces support for hnsw snapshots.
At start time the latest stored snapshot is read (if any), existing commit logs are sequentially processed and applied into the initial state.
Finally, the resulting state is persisted as a new snapshot, original snapshot and processed commit logs are removed because the latest snapshot captures all the changes up to this point.

NOTE: The changes are backward compatible but once applied it can not be downgraded as older versions are not able to read snapshots.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/vector/hnsw/startup.go
Comment (asdine) [inline adapters/repos/db/vector/hnsw/startup.go:167]: Not sure if it was by design, but it seems to be keeping the last two snapshots instead of one.
- the call to `getLastSnapshot` (L.62) makes sure there is only one snapshot left
- then in this condition we create a new snapshot if we found a commit log
- then in this line we remove the commit log: should we remove the previous snapshot as well?

**Comment by asdine** [comment]
After a few runs, with no data in the queue, we end up with two snapshots and an empty commit log.

image

Then every time we restart the server, a new snapshot is created (we can notice the `.tmp` file for a few seconds) and we end up with the state shown in the image above.

Do you think we can avoid doing this extra work if the commit log is empty?

**Comment by jeroiraz** [comment]
> After a few runs, with no data in the queue, we end up with two snapshots and an empty commit log.
> 
> image
> Then every time we restart the server, a new snapshot is created (we can notice the `.tmp` file for a few seconds) and we end up with the state shown in the image above.
> 
> Do you think we can avoid doing this extra work if the commit log is empty?

good, this would be a corner case, when shutting down without inserting any data. I think we could handle this case so to avoid unnecesary waits in such restart scenario

File: adapters/repos/db/vector/hnsw/startup.go
Comment (jeroiraz) [reply adapters/repos/db/vector/hnsw/startup.go:167]: it's by design so to be sure if something goes wrong we still have the previous consolidated one. Older snapshots will be erased when restarting...

Comment (jeroiraz) [reply adapters/repos/db/vector/hnsw/startup.go:167]: the older snapshot can be removed as soon as the new one is synced on disk

**Review by asdine — APPROVED** [review]
I ran the server a few times and tested it with no compression, pq, and bq. Recall is the same before and after restart.
For sq, the recall was also the same but I am not sure if it's because the benchmarker fails silently (since it's based on 1.24). cc @abdelr 

I recommend waiting for a review from @abdelr

**Pull request closed without merging**
